editor/src/components/editor/store/collaborative-editing.ts

```diff
@@ -0,0 +1,359 @@
+        change.projectFile.fileContents.parsed.type === 'PARSE_SUCCESS'
+      ) {
+        const parsedPart = change.projectFile.fileContents.parsed
+        let collabFile: CollabTextFile
+        if (projectContentsMap.has(change.fullPath)) {
+          collabFile = projectContentsMap.get(change.fullPath)!
+          // TODO Handle not text type
+        } else {
+          collabFile = new Y.Map()
+          collabFile.set('type', 'TEXT_FILE')
+          const topLevelElementsArray = new Y.Array<TopLevelElement>()
+          collabFile.set('topLevelElements', topLevelElementsArray)
+        }
+
+        mergeDoc.transact(() => {
+          if (!projectContentsMap.has(change.fullPath)) {
+            projectContentsMap.set(change.fullPath, collabFile)
+          }
+          synchroniseParseSuccessToCollabFile(parsedPart, collabFile)
+        })
+      }
+      break
+    case 'ENSURE_DIRECTORY_EXISTS':
+      // Not necessary to handle right now.
+      break
+    default:
+      assertNever(change)
+  }
+}
+
+export function updateCollaborativeProjectContents(
+  session: CollaborativeEditingSupportSession,
+  projectChanges: ProjectChanges,
+  filesModifiedByElsewhere: Array<string>,
+): void {
+  const projectContentsMap = session.projectContents
+  for (const change of projectChanges.fileChanges.collabProjectChanges) {
+    if (!filesModifiedByElsewhere.includes(change.fullPath)) {
+      applyFileChangeToMap(change, projectContentsMap, session.mergeDoc)
+    }
+  }
+}
+
+export function addHookForProjectChanges(
+  session: CollaborativeEditingSupportSession,
+  dispatch: EditorDispatch,
+): void {
+  session.projectContents.observeDeep((changeEvents) => {
+    // TODO Check that this is the array change before doing anything
+    for (const changeEvent of changeEvents) {
+      switch (changeEvent.path.length) {
+        case 0: {
+          if (changeEvent instanceof Y.YMapEvent) {
+            updateEntireProjectContents(session, changeEvent as Y.YMapEvent<any>, dispatch)
+          } else {
+            throw new Error(`Could not treat change event as Y.YMapEvent.`)
+          }
+          break
+        }
+        case 1: {
+          console.error(`Unhandled path length of 1: ${changeEvent.path}`)
+          break
+        }
+        case 2: {
+          const filePath = changeEvent.path[0] as string
+          if (changeEvent.path[1] !== 'topLevelElements') {
+            throw new Error(`Unexpected second part of change path: ${changeEvent.path[1]}`)
+          }
+          updateTopLevelElementsOfFile(session, filePath, changeEvent, dispatch)
+          break
+        }
+        default:
+          throw new Error(`Unexpected change path: ${JSON.stringify(changeEvent.path)}`)
+      }
+    }
+  })
+}
+
+export interface ArrayChanges {
+  updatesAt: Array<number>
+  deleteFrom: number | null
+}
+
+function updateEntireProjectContents(
+  session: CollaborativeEditingSupportSession,
```

Comment:
this does not seem to be used, can it be zapped?

Reply:
Fixed in 6d755a897.